File: cs/Problems/Day22.cs
namespace AOC2023.Problems;

public class Day22 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => CountSafelyDisintegratedBlocks(input);

    private int CountSafelyDisintegratedBlocks(ReadOnlySpan<char> input)
    {
        Span<Range> lineRanges = stackalloc Range[1500];
        int lines = input.Split(lineRanges, InputReader.NewLine);

        Span<Block> blocks = stackalloc Block[lines];

        for (int i = 0; i < lines; i++)
            blocks[i] = Block.Parse(input[lineRanges[i]]);

        Block.ApplyGravity(blocks);

        int safelyDisintegrated = 0;

        var blockQueue = new Queue<Block>(lines / 2);
        var falling = new HashSet<Block>(lines);
        var supportChain = SupportChain.BuildSupportChain(blocks);

        foreach (var disintegratingBlock in blocks)
        {
            blockQueue.Enqueue(disintegratingBlock);

            while (blockQueue.TryDequeue(out var block))
            {
                falling.Add(block);

                foreach (var above in supportChain.BlocksAbove[block])
                    if (supportChain.BlocksBelow[above].IsSubsetOf(falling))
                        blockQueue.Enqueue(above);
            }

            safelyDisintegrated += falling.Count - 1 == 0 ? 1 : 0;

            blockQueue.Clear();
            falling.Clear();
        }

        return safelyDisintegrated;
    }

    readonly record struct Block(LinearRange X, LinearRange Y, LinearRange Z)
    {
        public int Top => Z.End;

        public int Bottom => Z.Start;

        public static Block Parse(ReadOnlySpan<char> blockString)
        {
            Span<Range> numberRanges = stackalloc Range[6];
            blockString.SplitAny(numberRanges, [',', '~']);

            Span<int> numbers = stackalloc int[6];
            for (int i = 0; i < 6; i++)
                numbers[i] = int.Parse(blockString[numberRanges[i]]);

            return new Block(
                X: new LinearRange(numbers[0], numbers[3]),
                Y: new LinearRange(numbers[1], numbers[4]),
                Z: new LinearRange(numbers[2], numbers[5])
            );
        }

        public static void ApplyGravity(Span<Block> blocks)
        {
            MemoryExtensions.Sort(blocks, OrderByPosition);

            for (var i = 0; i < blocks.Length; i++)
            {
                int distance = 0;

                for (var j = 0; j < i; j++)
                    if (blocks[i].Intersects(blocks[j]))
                        distance = Math.Max(distance, blocks[j].Top);

                int fall = blocks[i].Bottom - distance - 1;
                blocks[i] = blocks[i] with { Z = blocks[i].Z.Reduce(fall) };
            }
        }

        public bool Intersects(Block other) => X.Intersects(other.X) && Y.Intersects(other.Y);

        static int OrderByPosition(Block a, Block b) => a.Bottom - b.Bottom;
    }

    readonly record struct LinearRange(int Start, int End)
    {
        public bool Intersects(LinearRange other) => Start <= other.End && other.Start <= End;

        public LinearRange Reduce(int amount) => new(Start - amount, End - amount);
    }

    readonly record struct SupportChain(Dictionary<Block, List<Block>> BlocksAbove, Dictionary<Block, HashSet<Block>> BlocksBelow)
    {
        public static SupportChain BuildSupportChain(ReadOnlySpan<Block> blocks)
        {
            var blocksAbove = new Dictionary<Block, List<Block>>(blocks.Length);
            var blocksBelow = new Dictionary<Block, HashSet<Block>>(blocks.Length);

            foreach (var block in blocks)
            {
                blocksAbove.Add(block, []);
                blocksBelow.Add(block, []);
            }

            for (var i = 0; i < blocks.Length; i++)
            {
                for (var j = i + 1; j < blocks.Length; j++)
                {
                    var zNeighbors = blocks[j].Bottom == 1 + blocks[i].Top;
                    if (zNeighbors && blocks[i].Intersects(blocks[j]))
                    {
                        blocksBelow[blocks[j]].Add(blocks[i]);
                        blocksAbove[blocks[i]].Add(blocks[j]);
                    }
                }
            }

            return new SupportChain(blocksAbove, blocksBelow);
        }
    }
}

File: go/problems/day22.go
package problems

import (
	"sort"
	"strconv"
	"strings"
)

type Day22 struct{}

func (d Day22) Solve(input string) int {
	return countSafelyDisintegratedBlocks(input)
}

func countSafelyDisintegratedBlocks(input string) int {
	lines := strings.Split(input, NewLine)
	blocks := make([]Block, len(lines))

	for i, line := range lines {
		blocks[i] = parseBlock(line)
	}

	applyGravity(blocks)

	safelyDisintegrated := 0
	blockQueue := make([]Block, 0)
	supportChain := buildSupportChain(blocks)

	for _, disintegratingBlock := range blocks {
		falling := make(map[Block]bool)
		blockQueue = append(blockQueue[:0], disintegratingBlock)

		for len(blockQueue) > 0 {
			block := blockQueue[0]
			blockQueue = blockQueue[1:]

			falling[block] = true

			for _, above := range supportChain.BlocksAbove[block] {
				if isSubset(falling, supportChain.BlocksBelow[above]) {
					blockQueue = append(blockQueue, above)
				}
			}
		}

		if len(falling)-1 == 0 {
			safelyDisintegrated++
		}
	}

	return safelyDisintegrated
}

type Block struct {
	X, Y, Z LinearRange
}

func (b Block) Bottom() int {
	return b.Z.Start
}

func (b Block) Top() int {
	return b.Z.End
}

func parseBlock(blockString string) Block {
	parts := strings.FieldsFunc(blockString, func(c rune) bool {
		return c == ',' || c == '~'
	})

	numbers := make([]int, len(parts))
	for i, part := range parts {
		num, _ := strconv.Atoi(part)
		numbers[i] = num
	}

	return Block{
		X: LinearRange{Start: numbers[0], End: numbers[3]},
		Y: LinearRange{Start: numbers[1], End: numbers[4]},
		Z: LinearRange{Start: numbers[2], End: numbers[5]},
	}
}

func applyGravity(blocks []Block) {
	sort.Slice(blocks, func(i, j int) bool {
		return blocks[i].Bottom() < blocks[j].Bottom()
	})

	for i := range blocks {
		distance := 0

		for j := 0; j < i; j++ {
			if blocks[i].Intersects(blocks[j]) {
				distance = max(distance, blocks[j].Top())
			}
		}

		fall := blocks[i].Bottom() - distance - 1
		blocks[i].Z = blocks[i].Z.Reduce(fall)
	}
}

func (b Block) Intersects(other Block) bool {
	return b.X.Intersects(other.X) && b.Y.Intersects(other.Y)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

type LinearRange struct {
	Start, End int
}

func (l LinearRange) Intersects(other LinearRange) bool {
	return l.Start <= other.End && other.Start <= l.End
}

func (l LinearRange) Reduce(amount int) LinearRange {
	return LinearRange{l.Start - amount, l.End - amount}
}

type SupportChain struct {
	BlocksAbove map[Block][]Block
	BlocksBelow map[Block]map[Block]bool
}

func buildSupportChain(blocks []Block) SupportChain {
	blocksAbove := make(map[Block][]Block)
	blocksBelow := make(map[Block]map[Block]bool)

	for _, block := range blocks {
		blocksAbove[block] = []Block{}
		blocksBelow[block] = make(map[Block]bool)
	}

	for i, block := range blocks {
		for j := i + 1; j < len(blocks); j++ {
			zNeighbors := blocks[j].Bottom() == 1+blocks[i].Top()
			if zNeighbors && block.Intersects(blocks[j]) {
				blocksBelow[blocks[j]][block] = true
				blocksAbove[block] = append(blocksAbove[block], blocks[j])
			}
		}
	}

	return SupportChain{BlocksAbove: blocksAbove, BlocksBelow: blocksBelow}
}

func isSubset(set1, set2 map[Block]bool) bool {
	for k := range set2 {
		if _, ok := set1[k]; !ok {
			return false
		}
	}
	return true
}
